README.txt
= README.txt

Release: $Release$

$Copyright$

http://cgialt.rubyforge.org/
http://rubyforge.org/projects/cgialt/
http://github.com/kwatch/cgialt/


==  About

'CGIAlt' is an alternative library of standard 'cgi.rb'.
It is written in pure Ruby but works faster than 'cgi.rb'.


== Features

* Compatible with 'cgi.rb'.
* Faster and more lightweight than 'cgi.rb'.
* Available to install with CGIExt (which is a re-implementation of 'cgi.rb'
  in C extension).
* FastCGI support


==  Install

Download cgialt-$Release$.tar.gz and install it according to the following:

    $ tar xzf cgialt-$Release$.tar.gz
    $ cd cgialt-$Release$/
    $ sudo ruby setup.rb

Or if you have installed RubyGems, you can install CGIAlt by 'gem' command.

    $ sudo gem install cgialt

It is recommended not to use RubyGems if you have to develop CGI program
because loading RubyGems is very heavy-weight for CGI program.


==  Usage

All you have to do is to require 'cgialt' instead of 'cgi', and you can use
CGI class which is compatible with 'cgi.rb'.

   require 'rubygems'   # if you have installed with RubyGems
   require 'cgialt'

If you want to use CGIAlt with FastCGI (fcgi.rb), require 'cgialt/fcgi'
instead of 'fcgi'.

   require 'cgialt'
   require 'cgialt/fcgi'

If you want require 'cgi' when CGIAlt is not installed, try the following.

   begin
     require 'cgialt'
   rescue LoadError
     require 'cgi'
   end
   begin
     reqiure 'cgialt/fcgi'
   rescue LoadError
     require 'fcgi'
   end

If you want to replace original 'cgi.rb' entirely by 'cgialt',
create 'cgi.rb' which requires 'cgialt' under proper directory such as
'/usr/local/lib/ruby/site_ruby/1.8'.

   $ sudo echo 'require "cgialt"' > /usr/local/lib/ruby/site_ruby/1.8/cgi.rb
   $ sudo echo 'require "cgialt/fcgi"' > /usr/local/lib/ruby/site_ruby/1.8/fcgi.rb

When $DEBUG is true or ENV['DEBUG'] is set, CGIAlt prints release number
to $stderr when loaded.

   $ DEBUG=1 ruby -e 'require "cgi"'
   *** CGIAlt: release $Release$


== Benchmark

Try 'bench.rb' included in CGIAlt archive.
The following is an example of benchmark.

    ## cgi.rb
    $ ruby -s bench.rb -N=1000
    *** 1000 times                         user    system     total       real
    ruby -e 'nil'                        0.0900    0.8300   11.5400 (  11.9631)
    ruby -e 'require "cgi"'              0.1000    1.2400   24.7000 (  25.2709)
    
    *** 100000 times                       user    system     total       real
    CGI#new (simple)                    20.3100    0.0300   20.3400 (  20.3770)
    CGI#new (complex)                   26.5800    0.0400   26.6200 (  26.6604)
    
    *** 1000000 times                      user    system     total       real
    CGI#header (simple)                 12.6700    0.0100   12.6800 (  12.7137)
    CGI#header (complex)                43.4600    0.0600   43.5200 (  43.5749)

    ## CGIAlt
    $ ruby -s bench.rb -N=1000 -cgialt
    *** CGIAlt: release 0.0.0
    *** 1000 times                         user    system     total       real
    ruby -e 'nil'                        0.0900    0.8000   11.5900 (  12.0581)
    ruby -e 'require "cgi"'              0.1000    1.2300   19.4800 (  20.0621)
    
    *** 100000 times                       user    system     total       real
    CGI#new (simple)                    14.5000    0.0300   14.5300 (  14.5594)
    CGI#new (complex)                   20.0700    0.0300   20.1000 (  20.1356)
    
    *** 1000000 times                      user    system     total       real
    CGI#header (simple)                  6.0400    0.0100    6.0500 (   6.0553)
    CGI#header (complex)                36.2200    0.0400   36.2600 (  36.3138)


It is good thing for performance to install CGIExt as well as CGIAlt.
The following is a benchmark example of using both CGIAlt and CGIExt.

    ## CGIAlt and CGIExt
    $ ruby -s bench.rb -N=1000 -cgialt -cgiext
    *** CGIAlt: release 0.0.0
    *** 1000 times                         user    system     total       real
    ruby -e 'nil'                        0.0900    0.8100   11.6600 (  12.1769)
    ruby -e 'require "cgi","cgiext"'     0.1000    1.2100   20.8400 (  21.5575)
    
    *** 100000 times                       user    system     total       real
    CGI#new (simple)                    12.4200    0.0400   12.4600 (  12.5207)
    CGI#new (complex)                   13.1600    0.0300   13.1900 (  13.2443)
    
    *** 1000000 times                      user    system     total       real
    CGI#header (simple)                  6.0300    0.0100    6.0400 (   6.0651)
    CGI#header (complex)                36.5400    0.0800   36.6200 (  37.0642)


The following is a summary of above benchmark results.

    Table 1. summary of benchmark
                                       cgi.rb      CGIAlt        CGIAlt+CGIExt
    ---------------------------------------------------------------------------
    require "cgi"         (x1000)       13.16      7.89 ( 67%)      9.18 ( 43%)
    CGI#new (simple)      (x100000)     20.34     14.53 ( 40%)     12.46 ( 63%)
    CGI#new (comple)      (x100000)     26.62     20.10 ( 32%)     13.19 (102%)
    CGI#header (simple)   (x1000000)    12.68      6.05 (110%)      6.04 (110%)
    CGI#header (complex)  (x1000000)    43.52     36.26 ( 20%)     36.62 ( 19%)


Another benchmark script 'bench.fcgi' is provided. It is for FastCGI.
The following is an example result of benchmark.

    Table 2. result of bench.fcgi
                                   Time taken for tests     Request per second
    --------------------------------------------------------------------------
    cgi    + fcgi                          16.686 [sec]        1198.61 [#/sec]
    cgialt + cgialt/fcgi                   15.562 [sec]        1285.18 [#/sec]
    cgialt + cgialt/fcgi + cgiext          15.310 [sec]        1306.34 [#/sec]


== License

Ruby's license


== Author

makoto kuwata <kwa(at)kuwata-lab.com>


== Bug reports

If you have bugs or questions, report them to kwa(at)kuwata-lab.com.


CHANGES.txt
## $Rev$
## $Release$
## $Copyright$

- release: 1.0.0
  date: 2008-11-09
  enhancements:
    - Ruby 1.9 support
    - Add tests for 'cgialt/util.rb'.

  changges:
    - CGI class now uses CGI#env_table() instead of using $CGI_ENV directly.
    - CGI class now uses CGI#stdinput() and CGI#stdoutput() instead of using $stdin and $stdout directly.
    - Refactor 'cgialt/fcgi/core.rb'.

  bugfixes:
    - CGI.parse() supports to parse 'a=10&b' (no '=').
    - CGI.parse() returned NoMethodError when QUERY_STRING was "&=10". Now fixed.
    - CGI.unescapeHTML() couldn't convert "&#32;". Now fixed.
    - Typo fixed in CGI::Cookie.parse().
    - Move RFC822_DAYS and RFC822_MONTHS from 'cgi/core.rb' to 'cgi/util.rb'.


- release: 0.0.2
  date: 2007-12-11
  enhancements:
    - FastCGI supported. If you required 'cgialt/fcgi' instead of 'fcgi',
      you can use FCGI.each_cgi() method.


- release: 0.0.1
  date: 2007-12-09


bench.rb
##
## $Rev$
## $Release: 0.0.0 $
## $Copyright$
##

##
## usage: ruby -s bench.rb [-cgialt] [-cgiext] [-rubygems] [-N=1000]
##

begin
  require($cgialt ? 'cgialt' : 'cgi')
rescue LoadError
  begin
    $:.unshift 'lib'
    require($cgialt ? 'cgialt' : 'cgi')
    $rubylib = 'lib'
  rescue LoadError
    require 'rubygems'
    require($cgialt ? 'cgialt' : 'cgi')
    $rubygems = true
  end
end

require 'benchmark'
require 'cgiext' if $cgiext
require 'rubygems' if $profile
require 'ruby-prof' if $profile

if defined?(CGI::RELEASE)
  puts "*** CGIAlt: release #{CGI::RELEASE}"
end

$N = ($N || 1000).to_i


class BenchmarkApplication


  QUERY_STRING = 'id=123&name=John+Smith&passwd=%40h%3D%7E%2F%5E%24%2F'
  #COOKIE_STRING = "name1=val1&val2&val3; name2=val2&%26%3C%3E%22;_session_id=abcdefg0123456789"
  COOKIE_STRING = '_session_id=e15eb25e3d8f3d51814024aab7ccbca0; dbx-pagemeta=grabit:0-|1-|2-|3-|4-|5-|6-|7-&advancedstuff:0-; dbx-postmeta=grabit:0+|1+|2+|3+|4+|5+|6+&advancedstuff:0-|1-|2-'
  #COOKIE_STRING = "name1=val1&val2&val3"
  SCRIPT_NAME  = '/cgi-bin/app/example/command.cgi'

  def initialize
    @request_method = 'GET'
    @query_string = QUERY_STRING
    @cookie_string = COOKIE_STRING
    @script_name = SCRIPT_NAME
  end

  def setup
    ENV['REQUEST_METHOD'] = @request_method
    ENV['QUERY_STRING']   = @query_string
    ENV['HTTP_COOKIE']    = @cookie_string
    ENV['SCRIPT_NAME']    = @script_name
  end


  def bench_invoke_ruby(ntimes)
    (ntimes/10).times do
      `ruby -e 'nil'`
      `ruby -e 'nil'`
      `ruby -e 'nil'`
      `ruby -e 'nil'`
      `ruby -e 'nil'`
      `ruby -e 'nil'`
      `ruby -e 'nil'`
      `ruby -e 'nil'`
      `ruby -e 'nil'`
      `ruby -e 'nil'`
    end
  end

  def self._def_bench_require(*names)
    if names.empty?
      method = 'bench_invoke_ruby'
      statement = 'nil'
    else
      method = "bench_require_" + names.join('_')
      statement = names.collect{|name| "require \"#{name}\""}.join('; ')
    end
    statement = "$:.unshift \"#{$rubylib}\"; " + statement if $rubylib
    command_opt = $rubygems ? '-rubygems' : ''
    s = ''
    s   << %Q|def #{method}(ntimes)\n|
    s   << %Q|  (ntimes / 10).times do\n|
    10.times do
      s << %Q|    `ruby #{command_opt} -e '#{statement}'`\n|
    end
    s   << %Q|  end\n|
    s   << %Q|end\n|
    return s
  end

  eval self._def_bench_require()
  eval self._def_bench_require('cgi')
  eval self._def_bench_require('cgi',    'cgiext')
  eval self._def_bench_require('cgialt')
  eval self._def_bench_require('cgialt', 'cgiext')

  def bench_cgi_new_simple(ntimes)
    ENV['QUERY_STRING'] = ''
    ENV['COOKIE_STRING'] = ''
    (ntimes / 10).times do
      CGI.new
      CGI.new
      CGI.new
      CGI.new
      CGI.new
      CGI.new
      CGI.new
      CGI.new
      CGI.new
      CGI.new
    end
  end

  def bench_cgi_new_complex(ntimes)
    ENV['QUERY_STRING'] =  @query_string
    ENV['COOKIE_STRING'] = @cookie_string
    (ntimes / 10).times do
      CGI.new
      CGI.new
      CGI.new
      CGI.new
      CGI.new
      CGI.new
      CGI.new
      CGI.new
      CGI.new
      CGI.new
    end
  end

  def bench_cgi_header_simple(ntimes)
    cgi = CGI.new
    (ntimes / 10).times do
      cgi.header("text/html; charset=utf8")
      cgi.header("text/html; charset=utf8")
      cgi.header("text/html; charset=utf8")
      cgi.header("text/html; charset=utf8")
      cgi.header("text/html; charset=utf8")
      cgi.header("text/html; charset=utf8")
      cgi.header("text/html; charset=utf8")
      cgi.header("text/html; charset=utf8")
      cgi.header("text/html; charset=utf8")
      cgi.header("text/html; charset=utf8")
    end
    return cgi.header("text/html; charset=utf8")
  end

  def bench_cgi_header_complex(ntimes)
    cgi = CGI.new
    options = {
      'type'     => 'text/xhtml',
      'charset'  => 'utf8',
      'length'   => 1024,
      'status'   => 'REDIRECT',
      'location' => 'http://www.example.com/',
      'server'   => 'webrick',
      'language' => 'ja',
      'Cache-Control' => 'none',
    }
    (ntimes / 10).times do
      cgi.header(options)
      cgi.header(options)
      cgi.header(options)
      cgi.header(options)
      cgi.header(options)
      cgi.header(options)
      cgi.header(options)
      cgi.header(options)
      cgi.header(options)
      cgi.header(options)
    end
    return cgi.header(options)
  end



  def execute

    method = "bench_require_#{$cgialt ? 'cgialt' : 'cgi'}#{$cgiext ? '_cgiext' : ''}"
    libname = "#{$cgialt ? 'cgialt' : 'cgi'}#{$cgiext ? '","cgiext': ''}"
    cmdopt = $rubygems ? '-rubygems ' : ''
    data = [
      [ 1,
        [ :bench_invoke_ruby, "ruby #{cmdopt}-e 'nil'" ],
        [ method.intern, "ruby #{cmdopt}-e 'require \"#{libname}\"'" ],
      ],
      [ 100,
        [ :bench_cgi_new_simple,  'CGI#new (simple)' ],
        [ :bench_cgi_new_complex, 'CGI#new (complex)' ],
      ],
      [ 1000,
        [ :bench_cgi_header_simple,  'CGI#header (simple)'  ],
        [ :bench_cgi_header_complex, 'CGI#header (complex)' ],
      ],
    ]

    nl = ''
    data.each do |list|
      factor = list.shift
      ntimes = factor * $N
      s = "#{ntimes} times"
      s << ' ' * (32 - s.length)
      header = "*** %s      user   system    total      real\n" % s
      format = "%8.3u %8.3y %8.3t %8.3r\n"
      print nl
      nl = "\n"
      Benchmark.benchmark(header, 38, format) do |job|
        list.each do |method, title|
          self.setup()
          output = nil
          if $profile
            result = RubyProf.profile do
              job.report(title) do
                output = self.__send__(method, ntimes)
              end
            end
            outfile = $o || "profile_#{method.to_s.sub(/bench_cgi_/,'')}.html"
            RubyProf::GraphHtmlPrinter.new(result).print(output='')
            File.open(outfile, 'w') {|f| f.write(output) }
          else
            job.report(title) do
              output = self.__send__(method, ntimes)
            end
          end
          if $DEBUG
            #puts "*** debug: "
            output.each {|line| p line }
          end
        end
      end
    end
  end

end


if $0 == __FILE__
  BenchmarkApplication.new.execute
end
